### ISSUE TITLE
Add error handling in dependency injection impl

### ISSUE THREAD

**Issue by sailingKieler** [langium/langium#463]
If
```javascript
value(injector)
```

of
https://github.com/langium/langium/blob/bc75d2866559224e96d632b50585624f1dffd803/packages/langium/src/dependency-injection.ts#L86

in
https://github.com/langium/langium/blob/bc75d2866559224e96d632b50585624f1dffd803/packages/langium/src/dependency-injection.ts#L83-L87

fails, `obj[prop] === __requested__` holds afterwards, and in a subsequent attempt the following error is thrown, which can be quite misleading:
https://github.com/langium/langium/blob/bc75d2866559224e96d632b50585624f1dffd803/packages/langium/src/dependency-injection.ts#L79-L81

I suggest to replace
https://github.com/langium/langium/blob/bc75d2866559224e96d632b50585624f1dffd803/packages/langium/src/dependency-injection.ts#L86

by
```javascript
    try {
        obj[prop] = (typeof value === 'function') ? value(injector) : _inject(value, injector);
    } catch (error) {
        obj[prop] = undefined;
        console.error(error);
        throw error;
    }
```

**Comment by Lotes** [comment]
How could a test for this look like? Shall I expect a "Cycle detected" when it is not fixed?

**Comment by sailingKieler** [comment]
I'm not sure whether a unit makes sense here.
If you can locally prove that this kind of misleading error report does not occur when a initialization error of an injected service occurs then it's fine.

If it had bee coded that way initially there wouldn't be a test on that, too.

**Comment by Lotes** [comment]
@sailingKieler question: -->

```typescript
        const value: Module<I, T[keyof T]> | ((injector: I) => T[keyof T]) = module[prop as keyof T];
        obj[prop] = __requested__;
        try {
            obj[prop] = (typeof value === 'function') ? value(injector) : _inject(value, injector);
        } catch (error) {
            obj[prop] = undefined; //DOES THIS LINE MAKE ANY SENSE WHEN WE THROW???
            console.error(error);
            throw error;
        }
        return obj[prop];
```

If we remove the commented line, the only benefit seems to be the error log... Is that worth it? Would be cool to have a scenario. I mean: is it even possible to reach the point where the undefined obj[prop] is used later? I think: NOPE

**Comment by sailingKieler** [comment]
> If we remove the commented line, the only benefit seems to be the error log... Is that worth it? Would be cool to have a scenario. I mean: is it even possible to reach the point where the undefined obj[prop] is used later? I think: NOPE

Yes, in my scenario (I don't remember all the details and the exact scenario anymore), the DI was asked again for that service after the initialization failed at the first attempt (I guess the dependency was required a second time somewhere else) and threw the `'Cycle detected. Please make ...` error that is basically a lie!

Thus, the current error indication is just wrong: if the initialisation of a service fails, it fails silently, and later on the misleading error is thrown and logged.

From the top of my head, I'm not sure whether the reset of `obj[prop]` to `undefined` is a must, at least it feels more correct to me as the DI's state would be left in a consistent shape, not in a kind of arbitrary state.

The more relevant part is the proper handling of potential initialization failures.
